TestP2PPeerReplicatorWithUpdateAndRestart is flaky

In the `Run tests job` action in the CI: https://github.com/sourcenetwork/defradb/actions/runs/4986293566/jobs/8926870794

```
=== FAIL: tests/integration/net/state/simple/peer_replicator TestP2PPeerReplicatorWithUpdateAndRestart (1.56s)
2023-05-16T00:04:15.783Z, INFO, tests.integration, , {"Database": "badger-in-memory"}
make[1]: *** [Makefile:165: test:names] Error 1
2023-05-16T00:04:15.791Z, INFO, tests.integration, Starting P2P node, {"P2P address": "/ip4/0.0.0.0/tcp/0"}
make: *** [Makefile:157: test:ci] Error 2
2023-05-16T00:04:15.805Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWLkRPR6VRR6RT3JZeKmFNiizF83P36dkNkjxqFjvRv1ok", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:15.809Z, INFO, node, Created LibP2P host, {"PeerId": "12D3KooWJgbjY8JooF1tj4wz9SePuwzVinXzfEGgvp1CgYz9cQRA", "Address": ["/ip4/0.0.0.0/tcp/0"]}
2023-05-16T00:04:15.813Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWEGYUECM3B65xN7EvLtt4bVhiSxCkbGRWmAnsZ4L8KVo8", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:15.821Z, INFO, net, Starting internal broadcaster for pubsub network
2023-05-16T00:04:15.830Z, INFO, tests.integration, Starting P2P node, {"P2P address": "/ip4/0.0.0.0/tcp/0"}
2023-05-16T00:04:15.849Z, INFO, node, Created LibP2P host, {"PeerId": "12D3KooWJu33tewtp8WeLPLP6f6LgAHKdg4hb5ZytS7b78y8Rjhf", "Address": ["/ip4/0.0.0.0/tcp/0"]}
2023-05-16T00:04:15.853Z, INFO, net, Starting internal broadcaster for pubsub network
2023-05-16T00:04:15.860Z, INFO, tests.integration, Starting P2P node, {"P2P address": "/ip4/0.0.0.0/tcp/0"}
2023-05-16T00:04:15.880Z, INFO, node, Created LibP2P host, {"PeerId": "12D3KooWDoHQyPzktNc9QmQ5Z3eR5HAdyVFN3dfRT4qAivRQdUsq", "Address": ["/ip4/0.0.0.0/tcp/0"]}
2023-05-16T00:04:15.885Z, INFO, net, Starting internal broadcaster for pubsub network
2023-05-16T00:04:16.002Z, INFO, tests.integration, Parsing bootstrap peers, {"Peers": "/ip4/10.1.0.121/tcp/40043/p2p/12D3KooWJgbjY8JooF1tj4wz9SePuwzVinXzfEGgvp1CgYz9cQRA"}
2023-05-16T00:04:16.003Z, INFO, tests.integration, Bootstrapping with peers, {"Addresses": [{"ID":"12D3KooWJgbjY8JooF1tj4wz9SePuwzVinXzfEGgvp1CgYz9cQRA","Addrs":["/ip4/10.1.0.121/tcp/40043"]}]}
2023-05-16T00:04:16.023Z, INFO, node, Connected, {"PeerID": "12D3KooWJgbjY8JooF1tj4wz9SePuwzVinXzfEGgvp1CgYz9cQRA"}
2023-05-16T00:04:16.029Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWJgbjY8JooF1tj4wz9SePuwzVinXzfEGgvp1CgYz9cQRA", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:16.036Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWJu33tewtp8WeLPLP6f6LgAHKdg4hb5ZytS7b78y8Rjhf", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:16.257Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWJgbjY8JooF1tj4wz9SePuwzVinXzfEGgvp1CgYz9cQRA", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:16.257Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWDoHQyPzktNc9QmQ5Z3eR5HAdyVFN3dfRT4qAivRQdUsq", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:16.274Z, INFO, merkleclock, Replacing DAG head, {"Old": "bafybeiajntql2kor3t73wxdtwfwv2ajyr23szk3ti47vqeporozm2xfviu", "CID": "bafybeieag7exqrxos6qd6jzveq7rcimi2h6vgwc2kmf5aow7ru7tjyxnte", "Height": 2}
2023-05-16T00:04:16.275Z, INFO, merkleclock, Replacing DAG head, {"Old": "bafybeiezptmalsierr4bbfnfecmzqwanjiufjbq7mwwnhm6zppqw7yixbm", "CID": "bafybeiddvpjup6743qjvhsu3bb7v25c3wmlsemigdpyqgwj7vrladf5gd4", "Height": 2}
2023-05-16T00:04:16.285Z, INFO, merkleclock, Replacing DAG head, {"Old": "bafybeiezptmalsierr4bbfnfecmzqwanjiufjbq7mwwnhm6zppqw7yixbm", "CID": "bafybeiddvpjup6743qjvhsu3bb7v25c3wmlsemigdpyqgwj7vrladf5gd4", "Height": 2}
2023-05-16T00:04:16.287Z, INFO, merkleclock, Replacing DAG head, {"Old": "bafybeiezptmalsierr4bbfnfecmzqwanjiufjbq7mwwnhm6zppqw7yixbm", "CID": "bafybeiddvpjup6743qjvhsu3bb7v25c3wmlsemigdpyqgwj7vrladf5gd4", "Height": 2}
2023-05-16T00:04:16.301Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWJu33tewtp8WeLPLP6f6LgAHKdg4hb5ZytS7b78y8Rjhf", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:16.304Z, INFO, merkleclock, Replacing DAG head, {"Old": "bafybeiajntql2kor3t73wxdtwfwv2ajyr23szk3ti47vqeporozm2xfviu", "CID": "bafybeieag7exqrxos6qd6jzveq7rcimi2h6vgwc2kmf5aow7ru7tjyxnte", "Height": 2}
2023-05-16T00:04:16.305Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWDoHQyPzktNc9QmQ5Z3eR5HAdyVFN3dfRT4qAivRQdUsq", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:16.306Z, INFO, merkleclock, Replacing DAG head, {"Old": "bafybeiajntql2kor3t73wxdtwfwv2ajyr23szk3ti47vqeporozm2xfviu", "CID": "bafybeieag7exqrxos6qd6jzveq7rcimi2h6vgwc2kmf5aow7ru7tjyxnte", "Height": 2}
2023-05-16T00:04:16.311Z, INFO, tests.integration, , {"RequestResults": [{"Age":60}]}
2023-05-16T00:04:16.312Z, INFO, tests.integration, , {"RequestResults": [{"Age":60}]}
2023-05-16T00:04:16.313Z, INFO, tests.integration, , {"RequestResults": [{"Age":60}]}
2023-05-16T00:04:16.313Z, INFO, tests.integration, , {"Database": "badger-file-system"}
2023-05-16T00:04:16.329Z, INFO, tests.integration, Starting P2P node, {"P2P address": "/ip4/0.0.0.0/tcp/0"}
2023-05-16T00:04:16.350Z, INFO, node, Created LibP2P host, {"PeerId": "12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB", "Address": ["/ip4/0.0.0.0/tcp/0"]}
2023-05-16T00:04:16.356Z, INFO, net, Starting internal broadcaster for pubsub network
2023-05-16T00:04:16.368Z, INFO, tests.integration, Starting P2P node, {"P2P address": "/ip4/0.0.0.0/tcp/0"}
2023-05-16T00:04:16.389Z, INFO, node, Created LibP2P host, {"PeerId": "12D3KooWL8XukQCxsMei2zcm9tapMxYJT5JaVVjkV957rVHZ31od", "Address": ["/ip4/0.0.0.0/tcp/0"]}
2023-05-16T00:04:16.395Z, INFO, net, Starting internal broadcaster for pubsub network
2023-05-16T00:04:16.408Z, INFO, tests.integration, Starting P2P node, {"P2P address": "/ip4/0.0.0.0/tcp/0"}
2023-05-16T00:04:16.430Z, INFO, node, Created LibP2P host, {"PeerId": "12D3KooWAoxsQYSHMr6juVShQtgQxqcfgPStxWQeCH2NBpC5YGnB", "Address": ["/ip4/0.0.0.0/tcp/0"]}
2023-05-16T00:04:16.436Z, INFO, net, Starting internal broadcaster for pubsub network
2023-05-16T00:04:16.556Z, INFO, tests.integration, Parsing bootstrap peers, {"Peers": "/ip4/10.1.0.121/tcp/35083/p2p/12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB"}
2023-05-16T00:04:16.556Z, INFO, tests.integration, Bootstrapping with peers, {"Addresses": [{"ID":"12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB","Addrs":["/ip4/10.1.0.121/tcp/35083"]}]}
2023-05-16T00:04:16.583Z, INFO, node, Connected, {"PeerID": "12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB"}
2023-05-16T00:04:16.584Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:16.590Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWL8XukQCxsMei2zcm9tapMxYJT5JaVVjkV957rVHZ31od", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:16.816Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:16.833Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "LEFT"}
2023-05-16T00:04:16.835Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "LEFT"}
2023-05-16T00:04:16.852Z, INFO, db, Closing DefraDB process...
2023-05-16T00:04:16.899Z	ERROR	peerstore/ds	pstoreds/keybook.go:81	error while updating pubkey in datastore for peer 12D3KooWAoxsQYSHMr6juVShQtgQxqcfgPStxWQeCH2NBpC5YGnB: datastore closed

2023-05-16T00:04:16.899Z	ERROR	peerstore/ds	pstoreds/keybook.go:81	error while updating pubkey in datastore for peer 12D3KooWAoxsQYSHMr6juVShQtgQxqcfgPStxWQeCH2NBpC5YGnB: datastore closed

2023-05-16T00:04:16.899Z, INFO, db, Successfully closed running process
2023-05-16T00:04:16.904Z, INFO, db, Closing DefraDB process...
2023-05-16T00:04:16.930Z, INFO, db, Successfully closed running process
2023-05-16T00:04:16.934Z, INFO, db, Closing DefraDB process...
2023-05-16T00:04:16.961Z, INFO, db, Successfully closed running process
2023-05-16T00:04:16.992Z, INFO, node, Created LibP2P host, {"PeerId": "12D3KooWAoxsQYSHMr6juVShQtgQxqcfgPStxWQeCH2NBpC5YGnB", "Address": ["/ip4/10.1.0.121/tcp/45539"]}
2023-05-16T00:04:17.002Z, INFO, net, Failure while reconnecting to a known peer, {"peer": "12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB", "error": "failed to dial 12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB:\n  * [/ip4/10.1.0.121/tcp/35083] dial tcp4 10.1.0.121:35083: connect: connection refused\n  * [/ip4/127.0.0.1/tcp/35083] dial tcp4 127.0.0.1:35083: connect: connection refused"}
2023-05-16T00:04:17.002Z, INFO, net, Starting internal broadcaster for pubsub network
2023-05-16T00:04:17.046Z, INFO, node, Created LibP2P host, {"PeerId": "12D3KooWL8XukQCxsMei2zcm9tapMxYJT5JaVVjkV957rVHZ31od", "Address": ["/ip4/10.1.0.121/tcp/38003"]}
2023-05-16T00:04:17.061Z, INFO, net, Failure while reconnecting to a known peer, {"peer": "12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB", "error": "failed to dial 12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB:\n  * [/ip4/127.0.0.1/tcp/35083] dial tcp4 127.0.0.1:35083: connect: connection refused\n  * [/ip4/10.1.0.121/tcp/35083] dial tcp4 10.1.0.121:35083: connect: connection refused"}
2023-05-16T00:04:17.061Z, INFO, net, Starting internal broadcaster for pubsub network
2023-05-16T00:04:17.121Z, INFO, node, Created LibP2P host, {"PeerId": "12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB", "Address": ["/ip4/10.1.0.121/tcp/35083"]}
2023-05-16T00:04:17.136Z, INFO, net, loaded replicators from datastore, {"Replicator": {"Info":{"ID":"12D3KooWAoxsQYSHMr6juVShQtgQxqcfgPStxWQeCH2NBpC5YGnB","Addrs":["/ip4/10.1.0.121/tcp/45539"]},"Schemas":["bafkreiddmpwv7legqpknh2kawgr3ev3peh3xehkzstt7fvocw2ubburb6a"]}}
2023-05-16T00:04:17.203Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:17.204Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWAoxsQYSHMr6juVShQtgQxqcfgPStxWQeCH2NBpC5YGnB", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:17.211Z, INFO, net, Starting internal broadcaster for pubsub network
2023-05-16T00:04:17.221Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWRUAhdUv4hF9WuMnV7z2PuJyKpBrXAQEFxGEZKgntu6bB", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:17.236Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWL8XukQCxsMei2zcm9tapMxYJT5JaVVjkV957rVHZ31od", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:17.268Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWL8XukQCxsMei2zcm9tapMxYJT5JaVVjkV957rVHZ31od", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:17.274Z, INFO, net, Received new pubsub event, {"SenderId": "12D3KooWAoxsQYSHMr6juVShQtgQxqcfgPStxWQeCH2NBpC5YGnB", "Topic": "bae-52b9170d-b77a-5887-b877-cbdbb99b009f", "Message": "JOINED"}
2023-05-16T00:04:17.323Z, INFO, merkleclock, Replacing DAG head, {"Old": "bafybeiajntql2kor3t73wxdtwfwv2ajyr23szk3ti47vqeporozm2xfviu", "CID": "bafybeieag7exqrxos6qd6jzveq7rcimi2h6vgwc2kmf5aow7ru7tjyxnte", "Height": 2}
2023-05-16T00:04:17.324Z, INFO, merkleclock, Replacing DAG head, {"Old": "bafybeiezptmalsierr4bbfnfecmzqwanjiufjbq7mwwnhm6zppqw7yixbm", "CID": "bafybeiddvpjup6743qjvhsu3bb7v25c3wmlsemigdpyqgwj7vrladf5gd4", "Height": 2}
2023-05-16T00:04:17.328Z, INFO, merkleclock, Replacing DAG head, {"Old": "bafybeiezptmalsierr4bbfnfecmzqwanjiufjbq7mwwnhm6zppqw7yixbm", "CID": "bafybeiddvpjup6743qjvhsu3bb7v25c3wmlsemigdpyqgwj7vrladf5gd4", "Height": 2}
2023-05-16T00:04:17.331Z, INFO, merkleclock, Replacing DAG head, {"Old": "bafybeiezptmalsierr4bbfnfecmzqwanjiufjbq7mwwnhm6zppqw7yixbm", "CID": "bafybeiddvpjup6743qjvhsu3bb7v25c3wmlsemigdpyqgwj7vrladf5gd4", "Height": 2}
2023-05-16T00:04:17.336Z, INFO, merkleclock, Replacing DAG head, {"Old": "bafybeiajntql2kor3t73wxdtwfwv2ajyr23szk3ti47vqeporozm2xfviu", "CID": "bafybeieag7exqrxos6qd6jzveq7rcimi2h6vgwc2kmf5aow7ru7tjyxnte", "Height": 2}
2023-05-16T00:04:17.340Z, INFO, tests.integration, , {"RequestResults": [{"Age":60}]}
2023-05-16T00:04:17.342Z, INFO, tests.integration, , {"RequestResults": [{"Age":21}]}
    utils2.go:1235: 
        	Error Trace:	/home/runner/work/defradb/defradb/tests/integration/utils2.go:1235
        	            				/home/runner/work/defradb/defradb/tests/integration/utils2.go:1047
        	            				/home/runner/work/defradb/defradb/tests/integration/utils2.go:400
        	            				/home/runner/work/defradb/defradb/tests/integration/utils2.go:283
        	            				/home/runner/work/defradb/defradb/tests/integration/net/state/simple/peer_replicator/with_update_restart_test.go:77
        	Error:      	Not equal: 
        	            	expected: 0x3c
        	            	actual  : 0x15
        	Test:       	TestP2PPeerReplicatorWithUpdateAndRestart
        	Messages:   	node: 1, doc: 0
2023-05-16T00:04:17.343Z, INFO, tests.integration, , {"RequestResults": [{"Age":60}]}
```

--- a/tests/integration/p2p.go
+++ b/tests/integration/p2p.go
@@ -159,13 +159,14 @@ func connectPeers(
 	// allowed to complete before documentation begins or it will not even try and sync it. So for now, we
 	// sleep a little.
 	time.Sleep(100 * time.Millisecond)
-	return setupPeerWaitSync(ctx, t, testCase, cfg, sourceNode, targetNode)
+	return setupPeerWaitSync(ctx, t, testCase, 0, cfg, sourceNode, targetNode)
 }
 
 func setupPeerWaitSync(
 	ctx context.Context,
 	t *testing.T,
 	testCase TestCase,
+	startIndex int,
 	cfg ConnectPeers,
 	sourceNode *node.Node,
 	targetNode *node.Node,
@@ -174,8 +175,8 @@ func setupPeerWaitSync(
 	sourceToTargetEvents := []int{0}
 	targetToSourceEvents := []int{0}
 	waitIndex := 0
-	for _, a := range testCase.Actions {
-		switch action := a.(type) {
+	for i := startIndex; i < len(testCase.Actions); i++ {
+		switch action := testCase.Actions[i].(type) {
 		case SubscribeToCollection:
 			if action.ExpectedError != "" {
 				// If the subscription action is expected to error, then we should do nothing here.
@@ -314,13 +315,14 @@ func configureReplicator(
 
 	_, err = sourceNode.Peer.SetReplicator(ctx, addr)
 	require.NoError(t, err)
-	return setupRepicatorWaitSync(ctx, t, testCase, cfg, sourceNode, targetNode)
+	return setupRepicatorWaitSync(ctx, t, testCase, 0, cfg, sourceNode, targetNode)
 }
 
 func setupRepicatorWaitSync(
 	ctx context.Context,
 	t *testing.T,
 	testCase TestCase,
+	startIndex int,
 	cfg ConfigureReplicator,
 	sourceNode *node.Node,
 	targetNode *node.Node,
@@ -330,8 +332,8 @@ func setupRepicatorWaitSync(
 	docIDsSyncedToSource := map[int]struct{}{}
 	waitIndex := 0
 	currentdocID := 0
-	for _, a := range testCase.Actions {
-		switch action := a.(type) {
+	for i := startIndex; i < len(testCase.Actions); i++ {
+		switch action := testCase.Actions[i].(type) {
 		case CreateDoc:
 			if !action.NodeID.HasValue() || action.NodeID.Value() == cfg.SourceNodeID {
 				docIDsSyncedToSource[currentdocID] = struct{}{}

--- a/tests/integration/utils2.go
+++ b/tests/integration/utils2.go
@@ -341,7 +341,7 @@ func executeTestCase(
 			// gracefully as part of the node closure.
 			syncChans = append(
 				syncChans,
-				restartNodes(ctx, t, testCase, dbt, nodes, dbPaths, nodeAddresses, nodeConfigs)...,
+				restartNodes(ctx, t, testCase, dbt, i, nodes, dbPaths, nodeAddresses, nodeConfigs)...,
 			)
 
 			// If the db was restarted we need to refresh the collection definitions as the old instances
@@ -604,6 +604,7 @@ func restartNodes(
 	t *testing.T,
 	testCase TestCase,
 	dbt DatabaseType,
+	actionIndex int,
 	nodes []*node.Node,
 	dbPaths []string,
 	nodeAddresses []string,
@@ -654,6 +655,20 @@ func restartNodes(
 		nodes[i] = n
 	}
 
+	// The index of the action after the last wait action before the current restart action.
+	// We wish to resume the wait clock from this point onwards.
+	waitGroupStartIndex := 0
+actionLoop:
+	for i := actionIndex; i >= 0; i-- {
+		switch testCase.Actions[i].(type) {
+		case WaitForSync:
+			// +1 as we do not wish to resume from the wait itself, but the next action
+			// following it. This may be the current restart action.
+			waitGroupStartIndex = i + 1
+			break actionLoop
+		}
+	}
+
 	syncChans := []chan struct{}{}
 	for _, tc := range testCase.Actions {
 		switch action := tc.(type) {
@@ -661,13 +676,13 @@ func restartNodes(
 			// Give the nodes a chance to connect to each other and learn about each other's subscribed topics.
 			time.Sleep(100 * time.Millisecond)
 			syncChans = append(syncChans, setupPeerWaitSync(
-				ctx, t, testCase, action, nodes[action.SourceNodeID], nodes[action.TargetNodeID],
+				ctx, t, testCase, waitGroupStartIndex, action, nodes[action.SourceNodeID], nodes[action.TargetNodeID],
 			))
 		case ConfigureReplicator:
 			// Give the nodes a chance to connect to each other and learn about each other's subscribed topics.
 			time.Sleep(100 * time.Millisecond)
 			syncChans = append(syncChans, setupRepicatorWaitSync(
-				ctx, t, testCase, action, nodes[action.SourceNodeID], nodes[action.TargetNodeID],
+				ctx, t, testCase, waitGroupStartIndex, action, nodes[action.SourceNodeID], nodes[action.TargetNodeID],
 			))
 		}
 	}

--- a/tests/integration/utils2.go
+++ b/tests/integration/utils2.go
@@ -658,18 +658,19 @@ func restartNodes(
 	for _, tc := range testCase.Actions {
 		switch action := tc.(type) {
 		case ConnectPeers:
+			// Give the nodes a chance to connect to each other and learn about each other's subscribed topics.
+			time.Sleep(100 * time.Millisecond)
 			syncChans = append(syncChans, setupPeerWaitSync(
 				ctx, t, testCase, action, nodes[action.SourceNodeID], nodes[action.TargetNodeID],
 			))
 		case ConfigureReplicator:
+			// Give the nodes a chance to connect to each other and learn about each other's subscribed topics.
+			time.Sleep(100 * time.Millisecond)
 			syncChans = append(syncChans, setupRepicatorWaitSync(
 				ctx, t, testCase, action, nodes[action.SourceNodeID], nodes[action.TargetNodeID],
 			))
 		}
 	}
-
-	// Give the nodes a chance to connect to each other and learn about each other's subscrivbed topics.
-	time.Sleep(100 * time.Millisecond)
 
 	return syncChans
 }

